anesthetic/utils.py

```diff
@@ -343,7 +343,7 @@ def sample_compression_1d(x, w=None, n=1000):
     if sum(w != 0) < n:
         i = w.index
     else:
-        i = np.random.choice(w.index, size=n, replace=False, p=w/w.sum())
+        i = np.random.choice(w.index, size=n, replace=False)
```

Comment:
This change seems to be doing exactly what I wanted. By turning this weighted subsample into a uniformly distributed subsample, it actually retains some points from the tails.

From what I can tell, this still works with the weight redistribution in the following lines. However, you are more familiar with this. Is there any danger to discarding the weights in this subsample generation?

Reply:
This is what this change looks like:
```python
ns = NestedSamples(data=data[['n_s', 'r', 'logr']], logL=data.logL, logL_birth=data.logL_birth, w=data.weight)
fig = plt.figure(figsize=(15, 5))
fig, axes = make_1d_axes(['n_s', 'r', 'logr'], fig=fig, ncols=3)
ns.plot_1d(axes, alpha=0.7, ncompress=1000 , label="ncompress=1000")
ns.plot_1d(axes, alpha=0.7, ncompress=50000, label="no compression")
axes['r'].set_xlim(0, 0.2)
axes['r'].legend()
```
![better_1d_compression](https://user-images.githubusercontent.com/18258042/85513881-5bcc0880-b5f3-11ea-8b13-75b0e5bb64d4.png)

Reply:
This is actually a very interesting and possibly extremely neat solution, as you are basing the binning on the nested sampling 'scan' rather than the posterior sample locations (and therefore get good tail coverage). The weights are still used and compressed correctly, but the location of the bins is far more sensible.

It's clear that this is better than what is currently being done, but I'm concerned that we're still seeing the compression level still affects some posteriors.

Taking a look at the code in sample_compression_1d, it looks like there is an edge-effect error which may be causing the behaviour we see in the middle plot, and may also be contributing to #93. I have just pushed a fix to this branch, and would appreciate your reproducing this plot on your system to see how things change.

If this does provide the correct behaviour, then there may be a corresponding better way of doing plot_2d.

Reply:
More importantly, what is the update to this plot with the new compression scheme?

Reply:
Here the updated plot:

```python
ns = NestedSamples(data=data[['n_s', 'r', 'logr']], logL=data.logL, logL_birth=data.logL_birth, weight=data.weight)
fig = plt.figure(figsize=(15, 5))
fig, axes = make_1d_axes(['n_s', 'r', 'logr'], fig=fig, ncols=3)
ns.plot_1d(axes, alpha=0.7, ncompress=1000 , label="ncompress=1000")
ns.plot_1d(axes, alpha=0.7, ncompress=5000 , label="ncompress=5000")
ns.plot_1d(axes, alpha=0.7, ncompress=50000, label="no compression")
axes['r'].set_xlim(0, 0.2)
axes['r'].legend()
```

![better_1d_compression2](https://user-images.githubusercontent.com/18258042/85539068-5f6c8900-b60d-11ea-8e91-6d532f3a7cdf.png)

It doesn't change much, but the difference to the full compression was decreased a little.

Reply:
If we were to be scientific, it would be good to have the plot with the new compression scheme, but with posterior rather than nested compression (this amounts to putting a `p=w/w.sum()` in the`np.random.choice`). I imagine the tail behaviour will be worse, but would like to see what the posterior bulk portion looks like. (It may be that the optimum is to have a combination of both).

Reply:
Here with `p=w/w.sum()` put back in:

![better_1d_compression_posteriorcompression](https://user-images.githubusercontent.com/18258042/85541497-f4708180-b60f-11ea-96b2-def4263a7822.png)

Reply:
OK, so the nested version looks like it's definitely better, and doesn't compromise things a great deal in the tails.

The only remaining question is whether for example in the ns and r case we want to plot that much of the tail (it's clear that we want it in the logr plot), but since the other two have high posterior compression I think the limits are now too wide. The role of the quantile portion in the original code (still in master) was to try and only plot the regions of interest. If you rein-state the code 
```python
i = ((x < quantile(x, 0.999, w)) & (x > quantile(x, 0.001, w))) | (p > 0.1)
```
 how do the plots above change?

Reply:
(but keep the nested binning rather than posterior binning)

Reply:
I would also be interested to know how this update affects #93.

Reply:
Updated plot using 
```python
i = ((x < quantile(x, 0.999, w)) & (x > quantile(x, 0.001, w))) | (p > 0.1)
```

![better_1d_compression_percentiles](https://user-images.githubusercontent.com/18258042/85543566-179c3080-b612-11ea-9ef6-5d13cab1c3e9.png)

Reply:
I don't like this because with this I have no chance of manually improving things.

On the other hand, it is easy to manually change limits, e.g.:
```python
axes['r'].set_xlim(0, 0.2)
```
as already done above. Note that the `r` plot is extreme because it's from a `logr` sampling run.

Reply:
> I would also be interested to know how this update affects #93.

Without the percentiles this would fix #93.

Reply:
> Without the percentiles this would fix #93.

￼This is very good.

> I don't like this because with this I have no chance of manually improving things.

I agree that there is definitely a hard-coded element to this, and one might often want to adjust the limits in other cases. 

However, as it stands this does not produce a very satisfactory plot by default, and would require manual adjustment (which would be impractical for large plots).

Is there a way that we could have the default bounds of the plot focusing on the posterior bulk, but still having the line in theory plotting the entire range?

Reply:
> Is there a way that we could have the default bounds of the plot focusing on the posterior bulk, but still having the line in theory plotting the entire range?

Introduce a `percentile` kwarg?

Reply:
I think the issue is that we currently use xmin and xmax in two ways, to both trim the plot bounds, and to adjust the kde at the edges. This is the source of #105, and possibly #94 and #95.

I imagine that there is a single PR that fixes all of these. The question is whether we make it this one, or find a reasonable half-way house to merge and close this.

Reply:
I already looked into a `percentile` kwarg, so I pushed it. Happy to revert if you want to leave that to another PR.

Tackling all of #94, #95 and #105 is more than I'd go for in this PR.

Reply:
Agreed